packages/open-normal-extension/open_normal.sh

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
+# SPDX-FileCopyrightText: 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+#
+# Chromium browser extension action script to open a URL in normal browser
+# Uses stdin and stdout for communicating with browser extension API
+# First four bytes tell the size of payload, and rest is the payload,
+# which is standard json data. Same format is used for input and output.
+
+# Configuration file which contains the givc parameters
+CFGF="/etc/open-normal-extension.cfg"
+
+# Get integer (ASCII code) from a byte
+function Ord {
+    printf "%d" "\"$1"
+}
+
+# Get byte from an integer (ASCII code)
+function Chr {
+    printf "%b" "$(printf "\\\\x%02x" "$1")"
+}
+
+# Send browser extension API message. First send four bytes of the length of
+# the payload and then the actual payload.
+function Msg {
+    local len b1 b2 b3 b4
+
+    len="${#1}"
+    b1="$(( len & 255 ))"
+    b2="$(( (len >> 8) & 255 ))"
+    b3="$(( (len >> 16) & 255 ))"
+    b4="$(( (len >> 24) & 255 ))"
+    Chr "$b1"
+    Chr "$b2"
+    Chr "$b3"
+    Chr "$b4"
+    printf "%s" "$1"
+}
+
+# Read the four bytes of length in the received API message
+LANG=C IFS= read -r -d '' -n 1 B1
+LANG=C IFS= read -r -d '' -n 1 B2
+LANG=C IFS= read -r -d '' -n 1 B3
+LANG=C IFS= read -r -d '' -n 1 B4
+
+# If read byte is zero, the corresponding variables are unset
+# Convert bytes to their ASCII values or zero if unset
+if [ -z "$B1" ]; then
+    B1=0
+else
+    B1="$(Ord "$B1")"
+fi
+if [ -z "$B2" ]; then
+    B2=0
+else
+    B2="$(Ord "$B2")"
+fi
+if [ -z "$B3" ]; then
+    B3=0
+else
+    B3="$(Ord "$B3")"
+fi
+if [ -z "$B4" ]; then
+    B4=0
+else
+    B4="$(Ord "$B4")"
+fi
+
+# Convert the four bytes to a integer length
+# Using multiply instead of bit shifing as bit shifts broke syntax highlighting
+LEN="$((B1+(B2*256)+(B3*65536)+(B4*16777216)))"
+
+# 2022 HTTP standard recommends supporting URLs up to 8000 characters
+# Generally URLs over 2000 characters will not work in the popular web browsers
+# So 8k should be plenty for the time being
+# If length is negative or zero or larger than 8k, then message is invalid
+if [ "$LEN" -le 0 ] || [ "$LEN" -gt 8192 ]; then
+    Msg "{\"status\":\"Failed to read parameters from API\"}"
+    exit 1
+fi
+
+# Read the json payload
+LANG=C IFS= read -r -d '' -n "$LEN" JSON
+# Remove json prefix
+PFX="{\"URL\":\""
+URL="${JSON##"$PFX"}"
+# Remove json suffix
+SFX="\"}"
+URL="${URL%%"$SFX"}"
```

Comment:
For future proofing could also use `jq`, like: `URL="$(dd count=$LEN status=none bs=1 | jq -r .URL)"` (jq would need the package dependency and correct path)

Reply:
I considered this a bit, and decided that having jq as dependency and having processing overhead of it is not justified here as input for the script is defined by another script in the same package.